### PR TITLE
Make WebAuthn required actions enabled by default

### LINKED ISSUE #12723: Make WebAuthn required actions enabled by default
### Description

It'd be great to have the `WebAuthn Register` and `WebAuthn Register Passwordless` enabled by default. 

### Discussion

https://github.com/keycloak/keycloak/discussions/9523

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by mabartos**
Closes #12723

@mposolda Could you please take a look at it? Thanks

### CODE REVIEW

File: services/src/main/java/org/keycloak/authentication/requiredactions/WebAuthnPasswordlessRegisterFactory.java [hunk 1]
```diff
@@ -20,13 +20,14 @@
...
 import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.UserModel;
 
 /**
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
  */
 public class WebAuthnPasswordlessRegisterFactory extends WebAuthnRegisterFactory {
 
-    public static final String PROVIDER_ID = "webauthn-register-passwordless";
+    public static final String PROVIDER_ID = UserModel.RequiredAction.WEBAUTHN_PASSWORDLESS_REGISTER.name();
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/authentication/requiredactions/WebAuthnPasswordlessRegisterFactory.java:30]: Is it possible to keep the Provider_ID same? I am not sure if change could affect migration (For example for these, who already have this required action enabled and registered and they migrate from previous version).

I can see for example in TermsAndConditions action that PROVIDER_ID does not need to be the name of the action on UserModel

File: services/src/main/java/org/keycloak/authentication/requiredactions/WebAuthnRegisterFactory.java [hunk 2]
```diff
@@ -35,7 +36,7 @@
 
 public class WebAuthnRegisterFactory implements RequiredActionFactory, DisplayTypeRequiredActionFactory, EnvironmentDependentProviderFactory {
 
-    public static final String PROVIDER_ID = "webauthn-register";
+    public static final String PROVIDER_ID = UserModel.RequiredAction.WEBAUTHN_REGISTER.name();
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/authentication/requiredactions/WebAuthnRegisterFactory.java:39]: Same comment as above

**Review by mposolda — CHANGES_REQUESTED** [review]
@mabartos Thanks, I think it would work fine if WebAuthn actions are registered by defaul in the realm. I've added 2 inline points. Besides that:

- I think the docs needs to be updated (AFAIK we have few places in the "WebAuthn" and "Authentication" sections, which mentions that WebAuthn actions need to be registered. This step can be updated (will be good to just update this step, but keep it, as people migrating their realms from previous versions won't have WebAuthn actions registered by default)

- WebAuthn tests are failing. I guess it is the regression of this? You are guru for these, so leaving to you to doublecheck :-)

**Comment by mabartos** [comment]
@mposolda Thanks for your comment and suggestions. I'll definitely add the missing stuff. 

Unfortunately, I've noticed the WebAuthn tests are not working in the GH actions due to some issues with the Chrome driver and its version. Related issue: #12621 

There's also a bug with the WebAuthn component with using it on the WF distribution. Related issue: #12762

File: server-spi-private/src/main/java/org/keycloak/models/utils/DefaultRequiredActions.java [hunk 3]
```diff
@@ -140,4 +142,40 @@ public static void addUpdateEmailAction(RealmModel realm){
...
+            webauthnRegister.setPriority(70);
+            realm.addRequiredActionProvider(webauthnRegister);
+        }
+    }
+
+    public static void addWebAuthnPasswordlessRegisterAction(RealmModel realm) {
+        final String PROVIDER_ID = "webauthn-register-passwordless";
+
+        final boolean isWebAuthnFeatureEnabled = Profile.isFeatureEnabled(Profile.Feature.WEB_AUTHN);
+        final boolean isRequiredActionActive = realm.getRequiredActionProviderByAlias(UserModel.RequiredAction.WEBAUTHN_PASSWORDLESS_REGISTER.name()) != null;
```
Comment (mposolda) [inline server-spi-private/src/main/java/org/keycloak/models/utils/DefaultRequiredActions.java:168]: Same comment as above.

File: server-spi/src/main/java/org/keycloak/models/UserModel.java [hunk 4]
```diff
@@ -311,7 +311,9 @@ enum RequiredAction {
         UPDATE_PASSWORD,
         TERMS_AND_CONDITIONS,
         VERIFY_PROFILE,
-        UPDATE_EMAIL
+        UPDATE_EMAIL,
+        WEBAUTHN_REGISTER,
```
Comment (mposolda) [inline server-spi/src/main/java/org/keycloak/models/UserModel.java:315]: Maybe there is no need to add these required actions to UserModel at all as they are not used anyway due the backwards compatibility (besides potentially incorrect usage, which I've pointed in the comments above). I think we already have few other required actions, which are not on UserModel. My vote is to revert this change and keep the UserModel class unchanged in this PR.

**Review by mposolda — CHANGES_REQUESTED** [review]
@mabartos Thanks for the heads-up. Added few minor inline comments.

Good luck with the fix of the WebAuthn on Wildfly. Hope you will be able to figure it.

**Comment by mabartos** [comment]
@mposolda Thanks for your comments. You're right. I need to verify all a little bit more :)

File: server-spi/src/main/java/org/keycloak/models/UserModel.java (hunk 4 above)
Comment (mabartos) [reply server-spi/src/main/java/org/keycloak/models/UserModel.java:315]: Agree.

File: server-spi-private/src/main/java/org/keycloak/models/utils/DefaultRequiredActions.java [hunk 5]
```diff
@@ -140,4 +142,40 @@ public static void addUpdateEmailAction(RealmModel realm){
             realm.addRequiredActionProvider(updateEmail);
         }
     }
+
+    public static void addWebAuthnRegisterAction(RealmModel realm) {
+        final String PROVIDER_ID = "webauthn-register";
+
+        final boolean isWebAuthnFeatureEnabled = Profile.isFeatureEnabled(Profile.Feature.WEB_AUTHN);
+        final boolean isRequiredActionActive = realm.getRequiredActionProviderByAlias(UserModel.RequiredAction.WEBAUTHN_REGISTER.name()) != null;
```
[the remark this answers is not in the document]
Comment (mabartos) [reply server-spi-private/src/main/java/org/keycloak/models/utils/DefaultRequiredActions.java:150]: +1

File: server-spi-private/src/main/java/org/keycloak/models/utils/DefaultRequiredActions.java (hunk 3 above)
Comment (mabartos) [reply server-spi-private/src/main/java/org/keycloak/models/utils/DefaultRequiredActions.java:168]: +1

**Comment by mabartos** [comment]
I've updated the commit. I can verify, that the `Webauthn Register` and `Webauthn Register Passwordless` required actions are enabled by default in the admin console. I'll execute the WebAuthn tests at least on my local machine.

**Comment by mabartos** [comment]
Rebased

**Review by mposolda — APPROVED** [review]
@mabartos Thanks for the updates!
